comet/comet.go

```diff
@@ -0,0 +1,467 @@
+		Height:   r.Height,
+		Index:    r.Index,
+		TxResult: r.Result,
+		Tx:       r.Tx,
+	}, nil
+}
+
+// TxSearch queries for multiple tx results. It returns a list of txs (max ?per_page_entries) and total count.
+// More: https://docs.cometbft.com/main/rpc/#/Tx/tx_search
+//
+// param pagePtr: 1-based page number, default (when pagePtr == nil) to 1
+// param perPagePtr: number of txs per page, default (when perPagePtr == nil) to 30
+// param orderBy: {"", "asc", "desc"}, default (when orderBy == "") to "asc"
+func (s *Tx) Search(
+	ctx *jsonrpctypes.Context,
+	query string,
+	prove bool,
+	pagePtr,
+	perPagePtr *int,
+	orderBy string,
+) (*rpctypes.ResultTxSearch, error) {
+	if prove {
+		return nil, errProvingNotSupported
+	}
+
+	q, err := bftquery.New(query)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse query: %w", err)
+	}
+
+	results, err := s.txstore.Search(ctx.Context(), q)
+	if err != nil {
+		return nil, fmt.Errorf("failed to search txs: %w", err)
+	}
+
+	// Sort results before pagination. Sort by height, then index.
+	switch orderBy {
+	case "desc":
+		sort.Slice(results, func(i, j int) bool {
+			if results[i].Height == results[j].Height {
+				return results[i].Index > results[j].Index
+			} else {
+				return results[i].Height > results[j].Height
+			}
+		})
+	case "asc", "": // Ascending order is the default.
+		sort.Slice(results, func(i, j int) bool {
+			if results[i].Height == results[j].Height {
+				return results[i].Index < results[j].Index
+			} else {
+				return results[i].Height < results[j].Height
+			}
+		})
+	default:
+		return nil, fmt.Errorf("expected order_by {asc, desc}, but got: '%s'", orderBy)
+	}
+
+	totalCount := len(results)
+	skipCount, returnedCount, err := paginate(pagePtr, perPagePtr, totalCount)
+	if err != nil {
+		return nil, err
+	}
+
+	apiResults := make([]*rpctypes.ResultTx, 0, returnedCount)
+	for i := skipCount; i < skipCount+returnedCount; i++ {
+		r := results[i]
+		apiResults = append(apiResults, &rpctypes.ResultTx{
+			Hash:     bfttypes.Tx(r.Tx).Hash(),
+			Height:   r.Height,
+			Index:    r.Index,
+			TxResult: r.Result,
+			Tx:       r.Tx,
+		})
+	}
+
+	return &rpctypes.ResultTxSearch{Txs: apiResults, TotalCount: totalCount}, nil
+}
+
+// Paginate calculates the skip count and actual page size for pagination based on the given parameters.
+// It takes the page number, page size, and total count as inputs and returns the skip count, actual page size, and any error encountered.
+// If the page number or page size is invalid, an error is returned.
+// The default page size is used if the page size is not provided or is out of range.
+// The total number of pages is calculated based on the total count and page size.
+// The skip count is calculated as (page - 1) * page size.
+// The actual page size is calculated as the minimum of the page size and the remaining count after skipping (boundary check).
+func paginate(pagePtr, pageSizePtr *int, totalCount int) (int, int, error) {
```

Comment:
int pointers here are a bit suspect - not much performance overhead to copying a new int but potentially subtle bugs emerging with modification of shared data.

This function is pretty susceptible to a bunch of off-by-one type errors. Suggestion: move to `utils/utils.go`, make public, add the following in utils_test.go

```go
func TestPaginate(t *testing.T) {
	type test struct {
		page,
		requestedPageSize,
		totalResults,
		expectedSkip,
		expectedPageSize int
		desc string
	}
	successes := []test{
		{1, 10, 0, 0, 0, "no results"},
		{1, 10, 1, 0, 1, "one result"},
		{1, 10, 10, 0, 10, "exactly one page"},
		{1, 10, 11, 0, 10, "one  non-returned result"},
		{2, 10, 11, 10, 1, "return single second-page result"},
		{2, 10, 20, 10, 10, "return 10 second-page results"},
		{16, 10, 200, 150, 10, "return intermediate page"},
	}

	failures := []test{
		{0, 10, 0, 0, 0, "pagenumber = 0"},
		{-1, 10, 100, 0, 0, "pagenumber < 0"},
	}

	for _, tt := range successes {
		skip, pageSize, err := comet.Paginate(&tt.page, &tt.requestedPageSize, tt.totalResults)
		require.Equal(t, tt.expectedSkip, skip)
		require.Equal(t, tt.expectedPageSize, pageSize)
		require.NoError(t, err)
	}

	for _, tt := range failures {
		_, _, err := comet.Paginate(&tt.page, &tt.requestedPageSize, tt.totalResults)
		require.Error(t, err)
	}
}
```

Reply:
FYI: the above does currently pass :)

Reply:
Amazing! Thanks so much for this test. We can add `comet/comet_pkg_test.go` with `package comet` to test the private `paginate` function.